### ISSUE TITLE
Model output meaningless boxes

### ISSUE THREAD

**Issue by bach05** [AMAP-ML/UniVG-R1#15]
Hello, 
I am trying the model but, while the reasoning process looks correct, the models fails on grounding it into the bounding box. I am attaching a minimal code to reproduce the issue. 

The output of the model is the following: 
`
Model output:
<think>To identify the differences between the two images, I will carefully compare each element in both images. First, I notice that the background and the hedgehog are similar in both images. Next, I focus on the objects on the hedgehog's back. In the first image, there is a red apple with a green leaf on top of the hedgehog. In the second image, the apple has changed to green. This change is the difference between the two images. To locate this difference within the relevant image, I will note the position of the apple in the first image.</think>
<answer>(500,270),(710,480)</answer>
`
Here the plotted box on the image: 

<img width="506" height="505" alt="Image" src="https://github.com/user-attachments/assets/cc702cbb-58dc-484f-b406-0180c59279a2" />

You can note that the input images are of shape 506x505, but the bounding box is completely out of this range. I noticed it happens also for different image resolutions. Am I missing something? Is the model supposed to work with images at a specific shape?

`
from transformers import Qwen2VLForConditionalGeneration, AutoProcessor
from qwen_vl_utils import process_vision_info
import torch

import re
from typing import List, Tuple, Union
from PIL import Image, ImageDraw

BBox = Tuple[int, int, int, int]

def parse_out_text(text: str) -> List[BBox]:
    """
    Parse VLM output formatted as one or more:
        <answer>(x1, y1),(x2,y2)</answer>
    or:
        <answer></answer>

    Returns:
        List of bounding boxes in xyxy format:
        [(x1, y1, x2, y2), ...]

    Notes:
    - Multiple <answer>...</answer> blocks are supported.
    - Empty <answer></answer> means no detection for that block.
    - If malformed answer blocks are found, they are skipped.
    """
    if text is None:
        return []

    # Extract all <answer>...</answer> blocks, allowing multiline contents
    answer_blocks = re.findall(r"<answer>(.*?)</answer>", text, flags=re.DOTALL | re.IGNORECASE)

    if not answer_blocks:
        return []

    bbox_pattern = re.compile(
        r"""
        \(\s*([-+]?\d*\.?\d+)\s*,\s*([-+]?\d*\.?\d+)\s*\)\s*,\s*
        \(\s*([-+]?\d*\.?\d+)\s*,\s*([-+]?\d*\.?\d+)\s*\)
        """,
        flags=re.VERBOSE,
    )

    bboxes: List[BBox] = []

    for block in answer_blocks:
        block_stripped = block.strip()

        # Explicit empty answer
        if block_stripped == "":
            continue

        match = bbox_pattern.fullmatch(block_stripped)
        if not match:
            # fallback: find bbox even if extra text exists
            match = bbox_pattern.search(block_stripped)

        if not match:
            continue

        try:
            x1, y1, x2, y2 = map(float, match.groups())
            bboxes.append((int(x1), int(y1), int(x2), int(y2)))
        except Exception:
            continue

    return bboxes

def plot_bboxes_on_image(
    image: Union[str, Image.Image],
    bboxes: List[BBox],
    output_path: str = None,
    width: int = 3,
) -> Image.Image:
    """
    Draw bounding boxes on an image.

    Args:
        image: path to image or PIL.Image
        bboxes: list of (x1, y1, x2, y2)
        output_path: optional path where to save the plotted image
        width: rectangle line width

    Returns:
        PIL.Image with drawn bounding boxes
    """
    if isinstance(image, str):
        img = Image.open(image).convert("RGB")
    else:
        img = image.convert("RGB")

    draw = ImageDraw.Draw(img)

    for i, (x1, y1, x2, y2) in enumerate(bboxes):
        draw.rectangle([x1, y1, x2, y2], outline="red", width=width)
        draw.text((x1, max(0, y1 - 12)), f"{i}", fill="red")

    if output_path is not None:
        img.save(output_path)

    return img

def main():
    max_pixels = 800 * 800
    model_path = "GD-ML/UniVG-R1"

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model = Qwen2VLForConditionalGeneration.from_pretrained(model_path).to(device)
    processor = AutoProcessor.from_pretrained(model_path, max_pixels=max_pixels)

    img0_path = "/media/iaslab/data_bacchin/img0.png"
    img1_path = "/media/iaslab/data_bacchin/img1.png"

    messages = [
        {
            "role": "user",
            "content": [
                {"type": "image", "image": img0_path},
                {"type": "image", "image": img1_path},
                {
                    "type": "text",
                    "text": (
                        "You are given two images. Spot the changes. "
                        "For each changed element, return its bounding box in the first image "
                        "as (x1, y1),(x2,y2). "
                        "First output the thinking process in <think> </think> tags. "
                        "Output only the result inside <answer></answer>."
                    ),
                },
            ],
        }
    ]

    text = processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
    image_inputs, video_inputs = process_vision_info(messages)
    inputs = processor(
        text=[text],
        images=image_inputs,
        videos=video_inputs,
        padding=True,
        return_tensors="pt",
    ).to(device)

    generated_ids = model.generate(**inputs, max_new_tokens=512)
    generated_ids_trimmed = [
        out_ids[len(in_ids):] for in_ids, out_ids in zip(inputs.input_ids, generated_ids)
    ]
    output_text = processor.batch_decode(
        generated_ids_trimmed,
        skip_special_tokens=True,
        clean_up_tokenization_spaces=False,
    )

    # batch_decode returns a list[str]
    pred_text = output_text[0]
    print("Model output:")
    print(pred_text)

    bboxes = parse_out_text(pred_text)
    print("Parsed bboxes:", bboxes)

    plotted = plot_bboxes_on_image(img0_path, bboxes, output_path="img0_with_bboxes.png")
    plotted.show()

if __name__ == "__main__":
    main()
`
Input images: 

<img width="506" height="505" alt="Image" src="https://github.com/user-attachments/assets/8cbcb6b6-e13d-4f53-b125-69b88aa85e38" />
<img width="506" height="505" alt="Image" src="https://github.com/user-attachments/assets/72001a35-2f78-4908-9640-a543c32adc92" />

**Comment by SuleBai** [comment]
Hi, thanks for your interest.

The coordinates output by our model are normalized to the [0, 1000] range, following the Qwen2-VL setting. Therefore, if you would like to visualize the predicted bounding box on the original image, you need to rescale the coordinates according to the image's actual width and height.

You can refer to the following example for the coordinate conversion:
```
from PIL import Image, ImageDraw

img = Image.open("img.png")
w, h = img.size

x1 = int(500 / 1000 * w)
y1 = int(270 / 1000 * h)
x2 = int(710 / 1000 * w)
y2 = int(480 / 1000 * h)

draw = ImageDraw.Draw(img)
draw.rectangle([x1, y1, x2, y2], outline="red", width=3)

img.save("output.png")
```
